text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+for experienced Ember developers when the computed property gets more
+complicated.
+
+Tracked properties have a feature called _autotrack_, where dependencies are
+automatically detected as they are used. This means that as long as all
+dependencies are marked as tracked, they will automatically be detected:
+
+```js
+import { tracked } from '@glimmer/tracking';
+
+class Person {
+  @tracked firstName = 'Tom';
+  @tracked lastName = 'Dale';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+This also allows us to opt out of tracking entirely, like if we know for
+instance that a given property is constant and will never change. In general,
+the idea is that _mutable_ properties should be marked as tracked, and
+_immutable_ properties should not.
+
+### Reducing Memory Consumption
+
+By default, computed properties cache their values. This is great when a
+computed property has to perform expensive work to produce its value, and that
+value gets used over and over again.
+
+But checking, populating, and invalidating this cache comes with its own
+overhead. Modern JavaScript VMs can produce highly optimized code, and in many
+cases the overhead of caching is greater than the cost of simply recomputing the
+value.
+
+Worse, cached computed property values cannot be freed by the garbage collector
+until the entire object is freed. Many computed properties are accessed only
+once, but because they cache by default, they take up valuable space on the heap
+for no benefit.
+
+For example, imagine this component that checks whether the `files` property is
+supported in input elements:
+
+```js
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+
+export default Component.extend({
+  inputElement: computed(function() {
+    return document.createElement('input');
+  }),
+
+  supportsFiles: computed('inputElement', function() {
+    return 'files' in this.inputElement;
+  }),
+
+  didInsertElement() {
+    if (this.supportsFiles) {
+      // do something
+    } else {
+      // do something else
+    }
+  },
+});
+```
+
+This component would create and retain an `HTMLInputElement` DOM node for the
+lifetime of the component, even though all we really want to cache is the
+Boolean value of whether the browser supports the `files` attribute.
+
+Particularly on mobile devices, where RAM is limited and often slow, we should
+be more conservative about our memory consumption. Tracked properties switch
+from an opt-out caching model to opt-in, allowing developers to err on the side
+of reduced memory usage, but easily enabling caching (a.k.a. memoization) if a
+property shows up as a bottleneck during profiling.
+
+## Detailed Design
+
+This RFC proposes adding the `tracked` decorator function, used to mark
+properties and getters as tracked:
+
+```ts
+const tracked: PropertyDecorator;
+```
+
+This new function will be exported from `@glimmer/tracking`. Revisiting our
```

Comment:
I would prefer if we could reexport from an `@ember` module. It is already hard to remember from where to import things, but adding an additional top-level module scope makes it even harder.

Reply:
The reasoning here is to keep `@tracked` compatible with both Glimmer.js _and_ Ember. This way, users will be able to write component libraries which are compatible with both.

We're working on another RFC at the moment which proposes adding GlimmerComponents to Ember, hopefully this will be submitted by the end of the week. It goes into more detail about this cross-compatibility.

Reply:
@pzuraq that's why I would propose reexporting so that it works with both paths. the Ember docs can then focus on e.g. `@ember/tracking` and people that want/need compat can know that both are supported

Reply:
That could be a possibility. We're proposing in the Glimmer Component RFC that `@glimmer/component` be a package independent of Ember.js and Glimmer.js, so the `@glimmer` namespace would be introduced then as well. I don't think a re-export makes as much sense for Glimmer Components, but it `@tracked` does not collide with any existing concepts in Ember so it could work.

Reply:
I came to post exactly this. Even as a long time ember user I found this very surprising and I think that @Turbo87's solution to re-export is a good compromise and will make this easier to teach.

Reply:
I don’t think having two places for it is a good idea. It doesn’t provide a huge benefit imo (editors like VS Code will auto import from the right place anyway) vs. the cost it introduces. 

Specifically, it’s a goal to have people write components that generally can run in Glimmer.js or Ember.js. Introducing an alternate module location just means that people will write components that aren’t portable to Glimmer for not very good reason. Putting the export in one place means people can copy and paste a component into a Glimmer.js app and have it “just work,” assuming it doesn’t really on obviously incompatible stuff like an app-specific service.

Reply:
I was initially against, but after reading these comments, I'm ok with the `@glimmer` namespace

Reply:
I vote for the re-export. Just as editors can auto-import, they can search & replace. I would definitely make it easier for the Ember developer (i.e. now we have to explain what "@glimmer" is), rather than the pure Glimmer developer. I'm sure they are way less in number

Reply:
FWIW, this isn't a new thing. Using things like `import { DEBUG } from '@glimmer/env'` is already public API and somewhat common (more so amongst addon authors I'd suspect).

Reply:
I will always vote for reducing the cognitive load on new Ember adopters, no matter how small. This is key for adoption, which we can agree Ember struggles with. If that glimmer API is most common among addon authors, then i'm fine with it.

Reply:
I would love to see the Ember project take a clear and consistent position on this, as it becomes a lot more of a sore spot once type information is involved. 

Another example of the same issue is the `Transition` type that some route hooks receive (a `router.js` concept that's exposed in Ember apps). Leaking one library's API through another makes questions like "what's the public API?" more difficult to answer.